### DISCUSSION TITLE
A general question about preconditioner

### DISCUSSION THREAD

**Discussion by alimostafavi24** [idaholab/moose#24282] [Q&A General] [answered]
Hello everyone,

I had a general question about the preconidtioner.

1. What is the difference between MOOSE and PETSC preconditioner? why we need to use them both?

2. Considering the below setup, what does the sub_pc mean? So we have a pc and sub_pc. What is the purpose of using both of them here. For example, what will happen if we do not use a sub_pc?

```
    petsc_options_iname = '-pc_type -pc_asm_overlap -sub_pc_type -sub_pc_factor_levels -sub_pc_factor_shift_type'
   petsc_options_value = 'asm      2               ilu          4                     NONZERO'
```

Thanks,
Ali

**Comment by GiudGiud** [comment, marked as answer]
Hello

A MOOSE 'preconditioner' is just setting the right parameters for a PETSc preconditioner.
For the more complex PETSc preconditioners, we have set up interfaces in MOOSE that can check parameters or make sure the PETSc parameters are passed in the desired order or with some more parameters tied to the size of the system that would not naturally be passed by a user. 

Many preconditioners are naturally composable and the sub_ option lets you specify options on each block (numerical)
For example see this page explaining the sub_ options for asm: https://petsc.org/release/manualpages/PC/PCGASM/

Guillaume

**Reply by alimostafavi24** [reply]
Thank you very much!

Ali
